### PR TITLE
Add Windows support

### PR DESCRIPTION
**Pull request by ljeromets**
#### Pull Request (PR) description
Hi,
This PR is based on the work done by @wardhus here https://github.com/voxpupuli/puppet-zabbix/pull/592

#### This Pull Request (PR) fixes the following issues: #175 
This PR should give the possibility to install zabbix-agent on Windows with chocolatey.

### CODE REVIEW

File: metadata.json [hunk 1]
```diff
@@ -146,10 +158,12 @@
       "operatingsystem": "Gentoo"
     },
     {
-      "operatingsystem": "Raspbian",
+      "operatingsystem": "windows",
```
Comment (bastelfreak) [inline metadata.json]: this needs to be rebased. Please don't drop Raspbian :)

File: manifests/agent.pp [hunk 2]
```diff
@@ -220,6 +223,8 @@
   $zabbix_version                                 = $zabbix::params::zabbix_version,
   $zabbix_package_state                           = $zabbix::params::zabbix_package_state,
   $zabbix_package_agent                           = $zabbix::params::zabbix_package_agent,
+  $zabbix_package_provider                        = $zabbix::params::zabbix_package_provider,
```
Comment (bastelfreak) [inline manifests/agent.pp]: Do we need to set ther provider explicity? Doesn't puppet pick the correct one on Windows? If it's required, please use a proper datatype like:
```suggestion
  Optional[String[1]] $zabbix_package_provider    = $zabbix::params::zabbix_package_provider,
```

File: manifests/agent.pp [hunk 3]
```diff
@@ -257,7 +262,7 @@
   Optional[Array] $zabbix_alias                   = $zabbix::params::agent_zabbix_alias,
   $timeout                                        = $zabbix::params::agent_timeout,
   $allowroot                                      = $zabbix::params::agent_allowroot,
-  $zabbix_user                                    = $zabbix::params::agent_zabbix_user,
+  Optional[String ] $zabbix_user                  = $zabbix::params::agent_zabbix_user,
```
Comment (bastelfreak) [inline manifests/agent.pp]:
```suggestion
  Optional[String[1]] $zabbix_user                = $zabbix::params::agent_zabbix_user,
```

File: manifests/agent.pp [hunk 4]
```diff
@@ -274,8 +279,8 @@
   $tlspskidentity                                 = $zabbix::params::agent_tlspskidentity,
   $tlsservercertissuer                            = $zabbix::params::agent_tlsservercertissuer,
   $tlsservercertsubject                           = $zabbix::params::agent_tlsservercertsubject,
-  String $agent_config_owner                      = $zabbix::params::agent_config_owner,
-  String $agent_config_group                      = $zabbix::params::agent_config_group,
+  Optional[String] $agent_config_owner            = $zabbix::params::agent_config_owner,
```
Comment (bastelfreak) [inline manifests/agent.pp]:
```suggestion
  Optional[String[1]] $agent_config_owner         = $zabbix::params::agent_config_owner,
```

File: manifests/agent.pp [hunk 5]
```diff
@@ -274,8 +279,8 @@
...
+  Optional[String] $agent_config_group            = $zabbix::params::agent_config_group,
```
Comment (bastelfreak) [inline manifests/agent.pp]:
```suggestion
  Optional[String[1]] $agent_config_group         = $zabbix::params::agent_config_group,
```

File: manifests/repo.pp [hunk 6]
```diff
@@ -159,6 +163,13 @@
         Apt::Source['zabbix'] -> Package<|tag == 'zabbix'|>
         Class['Apt::Update']  -> Package<|tag == 'zabbix'|>
       }
+      'windows' : {
+        if ($manage_choco == true) {
```
Comment (bastelfreak) [inline manifests/repo.pp]: this can be simplified:
```suggestion
        if $manage_choco {
```

File: manifests/repo.pp [hunk 7]
```diff
@@ -159,6 +163,13 @@
...
+          unless defined(Class['chocolatey']) {
+            include chocolatey
```
Comment (bastelfreak) [inline manifests/repo.pp]: at the moment I'm a bit unsure if we should keep this here of if chocolatey should be included in a profile

File: manifests/agent.pp (hunk 2 above)
Comment (ljeromets) [reply manifests/agent.pp]: hello,
It seems the default value for windows is "windows", MSI or Exe files. Source => https://puppet.com/docs/puppet/5.5/types/package.html#package-provider-windows.

I can add a condition like below ?
```
if $::kernel == 'windows' {
  package { $zabbix_package_agent:
    ensure   => $zabbix_package_state,
    require  => Class['zabbix::repo'],
    provider => $zabbix_package_provider,
    tag      => 'zabbix',
  }
 } else {
  # Installing the package
  package { $zabbix_package_agent:
    ensure   => $zabbix_package_state,
    require  => Class['zabbix::repo'],
    tag      => 'zabbix',
  }
```

File: manifests/repo.pp (hunk 7 above)
Comment (ljeromets) [reply manifests/repo.pp]: Ok, I can remove this call and remove the related entries from fixture.yml and metadata.json.

File: manifests/repo.pp (hunk 6 above)
Comment (ljeromets) [reply manifests/repo.pp]: I have removed it and added it in the agent.pp

File: manifests/agent.pp (hunk 2 above)
Comment (bastelfreak) [reply manifests/agent.pp]: I think the parameter is the right way to go. I was just hoping that puppet uses choco by default.

File: metadata.json (hunk 1 above)
Comment (ljeromets) [reply metadata.json]: hello it should be ok now

**Comment by ljeromets** [comment]
Hi @bastelfreak ,

I've added some units tests for windows but not sure they are really used.

The default value for `zabbix_version` is set to '3.4' but this version doesn't exists into chocolatey repo. ( https://chocolatey.org/packages/zabbix-agent#versionhistory)
Should we just add this limitation as a comment  or add another condition into the params.pp?

**Comment by bastelfreak** [comment]
@ljeromets thanks for all the updates. I propose that you add a selector block to the params.pp that sets `$zabbix_version` based on the kernel:

```patch
diff --git a/manifests/params.pp b/manifests/params.pp
index 501eff7..11c9eaa 100644
--- a/manifests/params.pp
+++ b/manifests/params.pp
@@ -119,6 +119,11 @@ class zabbix::params {
   }
 
   # Zabbix overall params. Is used by all components.
+  $zabbix_version = downcase($facts['kernel']) ? {
+    'windows' => '4.4.5',
+    default   => '3.4',
+  }
+
   $zabbix_package_state                     = 'present'
   $zabbix_proxy                             = 'localhost'
   $zabbix_proxy_ip                          = '127.0.0.1'
@@ -127,7 +132,6 @@ class zabbix::params {
   $zabbix_template_dir                      = '/etc/zabbix/imported_templates'
   $zabbix_timezone                          = 'Europe/Amsterdam'
   $zabbix_url                               = 'localhost'
-  $zabbix_version                           = '3.4'
   $zabbix_web                               = 'localhost'
   $zabbix_web_ip                            = '127.0.0.1'
   $manage_database                          = true
```

**Comment by bastelfreak** [comment]
Also while running the tests I noticed the following error message multiple times:
```
One or more the json documents could not be parsed. Run jgrep -v for to display documents
```

You can also see that in the Travis Ci output:
https://travis-ci.org/voxpupuli/puppet-zabbix/jobs/647484182?utm_medium=notification&utm_source=github_status

I assume some of the windows facts are broken :( I will have a closer look later at facterdb

**Comment by ljeromets** [comment]
@bastelfreak  I may have a idea about the json parsing issue.
```
PS C:\Users\Administrator> puppet facts | findstr operatingsystemrelease
    "operatingsystemrelease": "2012 R2",
```
I have update the metadata.json but I have now the error below:

> No facts were found in the FacterDB for Facter v3.14.0, using v3.8.0 instead

File: metadata.json [hunk 8]
```diff
@@ -151,6 +151,15 @@
         "9",
         "10"
       ]
+    },
+    {
+      "operatingsystem": "windows",
+      "operatingsystemrelease": [
+        "2008 R2",
```
Comment (bastelfreak) [inline metadata.json]: Please drop `2008 R2`, that one is EOL.

**Comment by ljeromets** [comment]
I would like help on the unit test ^^

File: manifests/agent.pp [hunk 9]
```diff
@@ -27,6 +27,13 @@
 # [*manage_repo*]
 #   When true, it will create repository for installing the agent.
 #
+# [*manage_choco*]
+#   When true on windows, it will use chocolatey to install the agent.
```
Comment (bastelfreak) [inline manifests/agent.pp]: Can you update this text? We should mention that https://forge.puppet.com/puppetlabs/chocolatey is required for this option

File: manifests/agent.pp [hunk 10]
```diff
@@ -286,8 +295,8 @@
   $tlspskidentity                                 = $zabbix::params::agent_tlspskidentity,
   $tlsservercertissuer                            = $zabbix::params::agent_tlsservercertissuer,
   $tlsservercertsubject                           = $zabbix::params::agent_tlsservercertsubject,
-  String $agent_config_owner                      = $zabbix::params::agent_config_owner,
-  String $agent_config_group                      = $zabbix::params::agent_config_group,
+  Optional[String[1]] $agent_config_owner            = $zabbix::params::agent_config_owner,
```
Comment (bastelfreak) [inline manifests/agent.pp]: can you please update the indentation here? all the `=` should be aligned

**Comment by bastelfreak** [comment]
@ljeromets thanks for the awesome work! I made two little inline comments. Can you please update the codebase and afterwards squash the commits down to one or a few? Afterwards we can merge and release this!

**Comment by ljeromets** [comment]
Hum, I don't know why the tests have failed.
I did another commit to trigger the checks, if they work I will clean back the git history.

**Comment by ljeromets** [comment]
Hello @bastelfreak ,
I will close this PR and open a fresh one, it is easier for me because I had too much issues with the git history.
But before that, can you help on the failling test?
```
  Warning: PostgresqlValidator.attempt_connection: Sleeping for 1 seconds
  Error: Unable to connect to PostgreSQL server! (:5432)
  Error: /Stage[main]/Postgresql::Server::Service/Postgresql_conn_validator[validate_service_is_running]/ensure: change from 'absent' to 'present' failed: Unable to connect to PostgreSQL server! (:5432)
```
source => https://travis-ci.org/voxpupuli/puppet-zabbix/jobs/653843178#L3867

**Comment by bastelfreak** [comment]
Thanks for the work! We merged it in https://github.com/voxpupuli/puppet-zabbix/pull/675

**Pull request closed without merging**